### ISSUE TITLE
Bundler should use already installed Gems, that satisfy Gemfile

### ISSUE THREAD

**Issue by bkabrda** [bundler/bundler#1964]
Hi,
I think that bundler should use already installed versions of Gems (doing bundle install), if they satisfy the Gemfile requirements. The current behaviour pollutes the system with new versions of Gems everytime I create a new Rails app. Not mentioning that I want to be able to use my systemwide Gems installed with RPM, that I know are supported by Fedora.

Thanks!

**Comment by akostadinov** [comment]
Your default use case assumes that you write an application that you deploy somewhere and whatever, you keep it in one place running and serving.
I think you didn't consider the use case where you have some tool that you run on many machines, maybe just once in the (virtual) machine life and you don't care so much about installed version but you care more about quick execution and avoiding issues with building native extensions on random machine configuration.
In such case a flag `--prefer-existing` would be extremely useful because it would allow bundle to use existing installed gems on the system when that's possible and install new gems only when needed. Would make support of multiple OS types and versions much easier. Instead of creating a lock file for each OS and version which might be a huge matrix and updates might come that will break existing lock files.
Why should be religious to not allowing such logic?

**Comment by sskylar** [comment]
Came here looking for the same thing as @akostadinov, a flag like `--prefer-existing` would be quite useful to avoid re-installing gems already installed at the system level.

**Comment by indirect** [comment]
@sskylar you can do that! The command is `bundle install --system`, which will use system-level gems if they are already installed. Since this ticket is closed, please open a new ticket if you have further issues or questions.

**Comment by sskylar** [comment]
@indirect Thanks for the reply! I'm aware of `bundle install --system`, but its not quite what I'm looking for. Let me try and explain this better...

We have a build server (Docker) that is loaded with a set of common system gems (Nokogiri, etc). This server is tasked with building various websites, each of which may use a subset of these system gems plus some additional gems on a per site basis.

Currently I'm running `bundle install --path .bundle` so each site has a separate cache of its required gems, however this also reinstalls gems that are already present on the system. This is a bigger pain for gems like Nokogiri that require building native extensions.

Ideally, something like `bundle install --path .bundle --prefer-existing` could support existing system gems where available, and only install new gems to `.bundle`. 

I've dug around here and it sounds like loading gems from multiple paths is something that used to be supported, but proved to be troublesome. Perhaps a solution to avoid that is to physically copy the gems from the system to `.bundle`, which certainly would be faster than downloading from the internet and rebuilding native extensions again.

I know its probably a stretch, so I won't open a new ticket unless this sounds like a feature you'd like to pursue.

**Comment by strzibny** [comment]
@sskylar It's sad to see how many people want this feature (and try to explain their use case here) yet the maintainers still don't see the usefulness. I almost wanted to fork bundler at some point because of it.

**Comment by indirect** [comment]
@sskylar thanks for explaining! can you help me understand what it is about this system that requires a hybrid setup using both --system and --path, rather than only using --system? Based on everything you have said so far, it seems like you could meet all of your goals by using system gems all the time.

@strzibny If you're not happy, you are of course absolutely welcome to fork Bundler at any time. Whether you are happy or not, passive aggressive comments attacking the judgement of the maintainers are not okay, and you need to stop.

**Comment by sskylar** [comment]
@indirect Thanks for hearing me out. In the setup I described, our system-installed gems are trusted gems that we've whitelisted and installed from rubygems.org. The sites we are building in Docker are user-supplied, and therefore the gems (and their sources) shouldn't be trusted by the system or be picked up by other users. For that reason, we're building each site to its own isolated `.bundle` file locally. However, since many common gems used by these sites are already installed at the system level, the `bundle install --path .bundle` is doing unnecessary work to re-download and build gems that are already present. While it works, the installation could be so much faster if it were able to read (but not write) from already installed system gems. 

Hope that makes sense :)

**Comment by indirect** [comment]
@sskylar yeah, I think that makes sense! It sounds like what you want to do is somewhat outside the scope of the core functionality Bundler is built around—we are aiming Bundler at end-user developers, not platform-building developers. That's why we removed the multiple paths feature long ago, and that's why we have avoided adding it back for so long.

I think your idea of copying the system gems into each site's path before installing is a reasonable one. If you want applications to each have their own separate set of gems, it even seems like it would probably be required—Bundler is not designed for multiple `bundle installs` run against the same directory at the same time, and will likely break in weird and hard to debug ways if you do that.

If there is consensus around what features make sense for platforms (vs users), maybe we can even roll those in to core someday. For now, we're focused squarely on prioritizing end user usability and expectations, which means sometimes deciding against things that would be useful for platforms. And if you do absolutely need to have Bundler use gems from multiple places at the same time, a fork to meet your needs as a platform is always an option. 

I hope that makes sense!

**Comment by chrismo** [comment]
this is a tangent, but maybe a workaround discussion that could become something for @sskylar. ignore with impunity!

i've been screwing around with a local setup that sounds similar to your needs (this [README](https://github.com/chrismo/bundle-bungler/blob/master/README.md) has my scenario/use-case), and came up with [this script](https://github.com/chrismo/bundle-bungler/blob/master/rbset_local_bundle.rb).

Basically it parses a local lockfile and tries to copy (rsync) all related files from a shared/global bundler path to the local project's bundle path. I say "tries" because the script is young and a few times I've found gaps, particularly with gems with native bits. This is a brittle because it's bypassing Bundler altogether to do the file syncing.

but, that said, it is frequently ~2-4x faster to rsync over an already built gem than have it pull everything clean from rubygems and build (esp. nokogiri, etc.) 

Once the copy is done, you should be able to `bundle install` to pick up anything that's unique to the local Gemfile.

A nice option for tinkering with this stuff with bundler is building out a plugin. (1.13, i think, has an even nicer plugin system, though what's there already is basic but works well). maybe this script of mine could be crafted into a plugin.

**Comment by sskylar** [comment]
Fair enough, thanks @indirect! 

@chrismo This looks very close to what we want, thanks for bringing it up. Peeking at your repo now 👀

**Comment by akostadinov** [comment]
I don't see any inherent contradiction between "end-users" and "integrators". It's just some more options that clueless people should not use. But this shouldn't be an argument to disallow these operations.
